util/test/computePerfStats

```diff
@@ -1,232 +1,302 @@
-            "head -c 1000 $test_output_file | tr -C '[:alnum:][:space:][:punct:]' '~'; ".
-            "tail -c 1000 $test_output_file | tr -C '[:alnum:][:space:][:punct:]' '~'; ".
-        "else ".
-            "cat $test_output_file | tr -C '[:alnum:][:space:][:punct:]' '~'; ".
-        "fi");
-    # we can't be sure if we end on a new line when limiting by characters
-    print "\n";
-    exit 1;
-}
-
-# remove white space surrounding a string
-sub trim($) {
-    my $string = shift;
-    $string =~ s/^\s+//;
-    $string =~ s/\s+$//;
-    return $string;
-}
+# COMPUTE PERFORMANCE STATISTTICS
+# This is used by sub_test if start_test is called with the -performance flag.
+# 
+# For every performance run on a particular file, this script maintains a .dat
+# by searching the output of the run for values of performance keys specified
+# in .perfkeys. It then writes these values to the .dat for every run.
+
+from __future__ import print_function
+import argparse;
+import os;
+import time;
+import re;
+import sys;
+import difflib;
+
+args = None
+
+def main():
+    parser = parser_setup()
+    global args
+    args = parser.parse_args()
+
+    setup()
+    (keys, verify_keys) = read_key_file()
+    valid_output = validate_output(verify_keys)
+
+    found_everything = True
+    if (valid_output):
+        create_data_file(keys)
+        log_timeouts(keys)
+        found_everything = find_keys(keys)
+    
+    if not (valid_output and found_everything) and not args.exec_time_out:
+        cleanup()
+
+
+def find_keys(keys):
+    found_everything = True
+    with open(data_file, "a") as file:
+        file.write("%s " % args.perfdate)
+        for key in keys:
+            sys.stdout.write("Looking for %s..." % key)
+            file.write("\t")
+            found = False
+            regex = r"(\s|\S)*" + re.escape(key) + r"\s*(\S*)"
+            # scan through output, looking for key
+            for line in test_output:
+                m = re.match(regex, line)
+                if m and not found:
+                    print("found it: %s" % m.group(2))
+                    file.write(m.group(2))
+                    found = True
+                    break
+
+            if not found:
+                file.write("-")
+                print("didn't find it")
+                found_everything = False
+        
+        file.write("\n")
+
+    return found_everything
+
+
+def log_timeouts(keys): # if we timed out
+    with open(errors_file, "w") as file:
+        file.write("appending %s\n" % data_file)
+
+    if args.exec_time_out:
+        print("ERROR")
+        with open(data_file, "a") as file:
+            file.write("#%s " % args.perfdate)
+            for key in keys:
+                file.write("\t-")
+            file.write(" ### EXECUTION TIMED OUT ###\n")
+        exit(1)
+
+
+def validate_output(verify_keys):
+    # read output from file
+    global test_output, test_output_raw
```

Comment:
Globals shouldn't be created inside of functions. These should either be defined outside of the function, or returned from this function.